### DISCUSSION TITLE
Get tags applied to a device via nbshell

### DISCUSSION THREAD

**Discussion by Mailstorm-ctrl** [netbox-community/netbox#16360] [Q&A] [answered]
Context: I'm trying to make a validator for a contact assignment that checks to see if the object the contact assignment is being made on has a specific tag applied. AKA: if tag not in parent.tags: fail

I can't seem to figure out how to do this without making an internal api call whcih I'd like to avoid because it seems unnecessary. In nbshell I can get the device object but when looking at the model I don't see any kind of tag property. So I'm not sure what I should be trying to do to accomplish this

**Comment by candlerb** [comment, marked as answer]
> AKA: if tag not in parent.tags: fail

It would have been helpful if you had shown the *actual* python code you tried, and the actual response.

`tags` is not a list or queryset, it's a database manager object. `tags.all()` will give you a queryset, which you *can* convert into a list if you want or construct an `in` query. Note that the tags themselves are are objects.

```
>>> i = IPAddress.objects.get(address="10.12.0.1/24")
>>> i.tags
<taggit.managers._TaggableManager object at 0x7bfdaf4e4cb0>
>>> i.tags.all()
<RestrictedQuerySet [<Tag: gateway>]>
>>> list(i.tags.all())
[<Tag: gateway>]
>>> Tag.objects.get(name="gateway") in i.tags.all()
True
```

If you are looking for a specific tag then there is a simpler approach: `tags.names()` gives you just the strings representing the tag names.

```
>>> i.tags.names()
<RestrictedQuerySet ['gateway']>
>>> list(i.tags.names())
['gateway']
>>> 'gateway' in i.tags.names()
True
```

**Reply by candlerb** [reply]
> In nbshell I can get the device object but when looking at the model I don't see any kind of tag property

```
>>> d = Device.objects.get(name="gw1")
>>> [m for m in dir(d) if "tag" in m]
['asset_tag', 'tagged_items', 'tags']
>>> d.tags
<taggit.managers._TaggableManager object at 0x7ef4a3e1b020>
>>> d.tags.all()
<RestrictedQuerySet []>

>>> help(Device)
...
 |  Method resolution order:
...
 |      netbox.models.features.TagsMixin
...
|  ----------------------------------------------------------------------
 |  Data and other attributes defined here:
...
 |
 |  tags = <taggit.managers._TaggableManager object>
 |
 |  ----------------------------------------------------------------------
```

**Reply by Mailstorm-ctrl** [reply]
I didn't show what I had written because I only got to the point of getting the device model with `Device.objects.get(pk=instance.object_id)`. Appending `.tags.all()` gets me the result I want...so thank you!

```
class monitoring_validator(CustomValidator):
    def validate(self, instance):
        if 12 in [tag.id for tag in instance._m2m_values['tags']] and 11 not in [tag.id for tag in Device.objects.get(pk=instance.object_id).tags.all()]:
            self.fail("The parent object must have the monitoring tag before assigning a contact tag to a contact!")
```

and then in the config
```
"tenancy.contactassignment": [
        "scripts.contactassignment-Validators.monitoring_validator"
    ]
```
